schemas/advanced-layouts.md

```diff
@@ -0,0 +1,120 @@
+
+Advanced layouts necessarily involve a form of logic to perform their data transformations. Many use-cases even necessitate interaction with block loading and storing mechanisms, such as traversal of a large `bytes` array stored across a chain of blocks. There is currently no formal requirement for where this logic resides, how it is loaded, or how it is associated with an IPLD Schema definition. This topic will continue to evolve. There is potential in the future for such logic to be embedded _in_ IPLD blocks themselves via WebAssembly, such that an advanced layout is represented both by the data and the logic required to read it. See [#130](https://github.com/ipld/specs/issues/130) for an early exploration of some of these themes.
+
+## Basic schema definition and use
+
+At its most basic, no properties are mandated for advanced layout. One may be defined simply with the keyword `advanced` in a similar manner to `type`:
+
+```ipldsch
+advanced ROT13
+```
+
+It may then be used as a `representation` for a `type`, from which we can infer the `kind` that we expect.
+
+```ipldsch
+type MyString string representation ROT13
+```
+
+Similarly, an advanced layout implementing a sharded `map` may be defined and used as:
+
+```ipldsch
+advanced ShardedMap
+
+type MyMap { String : Link } representation ShardedMap
+```
+
+From this usage, we may infer that `ShardedMap` can (1) present a familiar `map` kind interface and (2) store `link`s as values, referenced by standard data model `string`s. Other operating modes for `ShardedMap` may also be possible (it may be able to store other value kinds, or it may even be able to act as a `bytes` kind in spite of its name!).
+
+## Root node type definitions
+
+Advanced layouts are designed to abstract data that exists at the data model layer. As such, they may also dictate what they expect from the data that exists at the node their _root_ resides at.
+
+In the case of our `ROT13` `string` representation, we are likely to want to store this on the block as a `string` (i.e. this is a crude encryption mechanism, transforming `string` to `string`—realistic encryption mechanisms are likely to involve `bytes` and perhaps complex data structures to store encryption metadata).
+
+```ipldsch
+advanced ROT13 {
+  root String
+}
+
+type MyString string representation ROT13
+
+type Name {
+  firstname MyString
+  surname MyString
+}
+```
+
+A validator using our schema is now able to assert that it should find a `map` with two fields, `firstname` and `surname`, and, thanks to the `root` definition of `ROT13`, it may also assert that these two fields are of kind `string`.
+
+We may also introduce complex types as the root definition. For example, a `byte` representation that is a chain of blocks, each containing a section of `bytes` and a link to the next block:
+
+```ipldsch
+advanced ChainedBytes {
+  root Chunk
```

Comment:
I'm scared of this syntax in some way I can't quite put my finger on yet.

Glazing one's eyes over, the pattern of 'keyword TitleCase { linebreak fieldname TypeName }` looks like -- well, `fieldname` isn't a fieldname, is it.  This isn't a struct.

I wonder if some more visual distinctiveness wouldn't be a pleasant idea here.  I'm worried that someone might skim this and infer that there's literally a field named "root" somewhere in the data, which would be incorrect.

(I don't have a suggestion at the moment, either, sorry.  Just thinking outloud here.)

Reply:
(I know, I've also drafted sketches with roughly this syntax before.  I'm not sure why I'm more triggered now.  Possibly because in other drafts, there have been more keywords (second word of line being lowercase rather than titlecase) and other errata that made it more clearly not a struct.  Possibly because the brainmeats are just more critical when someone else does it :))

Reply:
Ahh, so this kind of thinking is what lead you to put a `|` in front of union members, _(personal `*click*` moment)_. Remember that your `representation {}` block does the same thing as this, `discriminatorKey` being a glaring example.
I have no strong feelings either way so I'll wait for suggestions.

Reply:
Yeah.

Worth noting perhaps that the syntax for the representation block has also drifted back and forth a bit over time.  Earlier drafts had `key="value"` pairs -- thus feeling more like value assignment, and distinct from "fieldname TypeName" declarations -- and I think might've even used parens rather than squirrely braces.  They were originally even one-liners, I think... but started looking too long that way.

Reply:
I find it OK having the first keyword (i.e. `advanced`, `struct`) indicating what the key value pairs inside mean. Whether they are fixed parameters of definitions. I agree the readability is more important then writability as your write once, read often.

Though if i think about writing a schema, I would learn what the different keywords `advanced` and `struct` mean and what they are doing. I wouldn't need to remember which syntax goes with them, I just need to remember a single concept "key-value pairs are represented as `key <space> value` in ipldsch.

Though we already have a different syntax for `union`s and `enum`s, which breaks the "single concept" argument.

When going for a different syntax `=` would make sense as it is an assignment. Though I'm not sure if I like this myself:

```
advanced ROT13 {
  = root String
}
```

Reply:
Or a `.`? That has "property" connotations:

```
advanced ROTX {
  .characters Int
}
```

I find it all a bit ugly and don't really need the `|` in unions or for this to be different. They're all struct-style definitions anyway. But that's a weakly held opinion.

Reply:
I think I withdraw most of my concerns here.  I'm much less phased when adding in other examples like adding the `param name "value"` triples mentioned in a subsequent PR.  It's mostly the specifics of "`Chunk`" in `root Chunk` being both a TypeName but having radically different semantics than any other appearance of a TypeName that gives me pause... which is maybe an issue better regarded on its own.